### PR TITLE
Initial WS Server

### PR DESCRIPTION
**Pull request by eabz**
Reference: #740, #499 #242 
@kyuupichan 
What do you think? Is very elemental, but is working. 
At least is working with my app

### CODE REVIEW

**Comment by erasmospunk** [comment]
Is it possible to add some tests?

**Comment by eabz** [comment]
@erasmospunk 
What kind of tests?

File: .travis.yml [hunk 1]
```diff
@@ -26,7 +26,9 @@ install:
   - pip install pytest-asyncio
   - pip install pytest-cov
   - pip install Sphinx
-# hashes
+  - pip install websockets
+  - pip install jsonrpcserver
```
[the remark this answers is not in the document]
Comment (eabz) [reply .travis.yml]: @kyuupichan Removed jsonrpcserver

**Comment by eabz** [comment]
@kyuupichan 
Remove jsonrpcserver and revert to websockets library.
Also, using ElectrumX all methods are available on websockets

File: electrumx/server/session.py [hunk 2]
```diff
@@ -142,6 +143,39 @@ def __init__(self, env, db, bp, daemon, mempool, shutdown_event):
         LocalRPC.request_handlers = {cmd: getattr(self, 'rpc_' + cmd)
                                      for cmd in cmds}
 
+    async def _start_ws_server(self, *args, **kw_args):
+        instance = ElectrumX(self, self.db, self.mempool, self.peer_mgr, 'TCP')
+
+        @asyncio.coroutine
```
Comment (kyuupichan) [inline electrumx/server/session.py]: The decorator is not necessary; it appears nowhere else in ElectrumX.

File: electrumx/server/env.py [hunk 3]
```diff
@@ -48,6 +48,7 @@ def __init__(self, coin=None):
         self.reorg_limit = self.integer('REORG_LIMIT', self.coin.REORG_LIMIT)
         # Server stuff
         self.tcp_port = self.integer('TCP_PORT', None)
+        self.websocket_port = self.integer('WEBSOCKET_PORT', None)
```
Comment (kyuupichan) [inline electrumx/server/env.py]: It's not clear if this is SSL or plain text communication.

Can we support both?  Perhaps with env vars WS_PORT and WSS_PORT?

**Comment by kyuupichan** [comment]
This is looking promising, thank you; give me a bit of time to review.  Please also see the 2 comments I just left.

I don't have client code to test with - can you confirm it works for you?

**Comment by kyuupichan** [comment]
Also please update the documentation for the envvar(s) you add.

File: electrumx/server/session.py (hunk 2 above)
Comment (eabz) [reply electrumx/server/session.py]: Removed

File: electrumx/server/env.py (hunk 3 above)
Comment (eabz) [reply electrumx/server/env.py]: @kyuupichan I think the best option is check for SSL_CERTFILE and SSL_KEY, if those are specified, ElectrumX will serve a Secure WebSocket

Comment (eabz) [reply electrumx/server/env.py]: Updated

**Comment by eabz** [comment]
@kyuupichan 
Added Secure WebSocket support and docs.
The Server is working, I am testing it against an app created on Ionic. 
I am still testing the WSS.

**Comment by eabz** [comment]
@kyuupichan 
Hello, I have tested it all weekend and is working. 
Currently supports:
- Websockets ("ws://")
- Secure websockets ("wss://")
- Version handshake
- Single querys
- Batches.

**Comment by kyuupichan** [comment]
Thanks for confirming.  Give me some time to review please.

**Comment by eabz** [comment]
@kyuupichan 
Hello, do you get a chance to review the PR?

**Comment by kyuupichan** [comment]
Thanks for this.  I've just taken a look.  The fact you have to worry anything about JSON means that it's not being abstracted correctly.  I think I need to rework the session class in aiorpcX, then integrating websockets should be easy.

**Comment by eabz** [comment]
Thanks @kyuupichan we are currently reworking our backend and we may don't use electrumx anymore, if you can take it from here it will be ok, but I am not working on it further. If this is a good start for you is perfect, but if don't, you can close it.

**Comment by kyuupichan** [comment]
I'll be making the session class independent of asyncio protocols; then it can simply having a transport.  That can be an asyncio one (suitably inherited) or it can be a websocket one (using the package).  Then in ElectrumX it will just be a simple change to add websocket support once I update it for the new interface.

I may copy some of your documentation for the new env vars, thanks.

**Comment by kyuupichan** [comment]
I just implemented websockets in https://github.com/kyuupichan/electrumx/commit/a6bfc032674aac71eeffea3457ad10c456c1baa0

See https://electrumx.readthedocs.io/en/latest/environment.html#services

Note how none of the server code really needed to change

**Pull request closed without merging**